### PR TITLE
Add AspaceClient service

### PR DESCRIPTION
**Pull request by corylown**

### CODE REVIEW

File: spec/services/aspace_client_spec.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  let(:client) { described_class.new(url:, user:, password:) }
+
+  before do
+    # Authentication request
+    stub_request(:post,
+                 "#{url}/users/#{user}/login?password=#{password}").to_return(body: { session: 'token1' }.to_json)
+  end
+
+  describe '#initialize' do
+    it 'fails without the url arg' do
```
Comment (jcoyne) [inline spec/services/aspace_client_spec.rb:18]: can you move any "without..." from the "it" to a "context" block?

File: spec/services/aspace_client_spec.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    it 'fails without the password arg' do
+      expect do
+        described_class.new(url: 'http://example.com:8089', user: 'user', password: nil)
+      end.to raise_error(ArgumentError)
+    end
+  end
+
+  describe '#repositories' do
+    it 'sends an authenticated request with correct auth header to the repositories address' do
```
Comment (jcoyne) [inline spec/services/aspace_client_spec.rb:36]: can you move any "with..." from the "it" to a "context" block?

File: spec/services/aspace_client_spec.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      stub_request(:get, "#{url}/repositories").to_return(body: {}.to_json)
+      client.repositories
+      expect(WebMock).to have_requested(:get, "#{url}/repositories")
+                     .with(headers: { 'X-ArchivesSpace-Session' => 'token1' }).once
+    end
+  end
+
+  describe '#resource_description' do
+    it 'fails if you do not pass resource_uri' do
```
Comment (jcoyne) [inline spec/services/aspace_client_spec.rb:45]: can you move any "if..." from the "it" to a "context" block?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: spec/services/aspace_client_spec.rb (hunk 2 above)
Comment (corylown) [reply spec/services/aspace_client_spec.rb:42]: Moved. Thank you.

File: app/services/aspace_client.rb [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    response = authenticated_get('repositories')
+
+    JSON.parse(response)
+  end
+
+  # Get the EAD XML representation of a resource
+  # See https://archivesspace.github.io/archivesspace/api/#get-an-ead-representation-of-a-resource
+  # @example client.resource_description('/repositories/2/resources/5363')
+  # @param [String] resource_uri the URI of the resource in ArchivesSpace
+  def resource_description(resource_uri)
```
Comment (marlo-longley) [inline app/services/aspace_client.rb:35]: This is super minor -- but the only thing I find slightly inconsistent in this class is that for this method, we expect the constructed url to be passed as an argument. Whereas further down in `published_resource_uris`, we accept `repository_id` as its own argument -- also in the `AspaceQuery` class we also take `repository_id` and  `resource_id` as distinct arguments and then construct the URL internally in the methods.

Comment (corylown) [reply app/services/aspace_client.rb:35]: There's a method to my madness: the `resource_description` method will be called with data from the API response which is already in the form we need `/repositories/2/resources/5363`. In the other cases we're starting with more limited information. 🤷 If it bugs us enough we can refactor.

**Pull request merged**